src/cuda/mod.rs

```diff
@@ -0,0 +1,437 @@
+            err
+        })?;
+        let stream = Stream::new(StreamFlags::NON_BLOCKING, None).map_err(|err| {
+            Self::pop_context();
+            err
+        })?;
+        let prog = Program {
+            module,
+            stream,
+            device_name: device.name(),
+            context: device.context.clone(),
+        };
+        Self::pop_context();
+        Ok(prog)
+    }
+
+    /// Creates a program for a specific device from a compiled CUDA binary.
+    pub fn from_bytes(device: &Device, bytes: &[u8]) -> GPUResult<Program> {
+        rustacuda::context::CurrentContext::set_current(&device.context)?;
+        let module = rustacuda::module::Module::load_from_bytes(bytes).map_err(|err| {
+            Self::pop_context();
+            err
+        })?;
+        let stream = Stream::new(StreamFlags::NON_BLOCKING, None).map_err(|err| {
+            Self::pop_context();
+            err
+        })?;
+        let prog = Program {
+            module,
+            stream,
+            device_name: device.name(),
+            context: device.context.clone(),
+        };
+        Self::pop_context();
+        Ok(prog)
+    }
+
+    /// Creates a new buffer that can be used for input/output with the GPU.
+    ///
+    /// The `length` is the number of elements to create.
+    ///
+    /// It is usually used to create buffers that are initialized by the GPU. If you want to
+    /// directly transfer data from the host to the GPU, you would use the safe
+    /// [`Program::create_buffer_from_slice`] instead.
+    ///
+    /// ### Safety
+    ///
+    /// The buffer needs to be initalized (by the host with [`Program::write_from_buffer`]) or by
+    /// the GPU) before it can be read via [`Program::read_into_buffer`].
+    pub unsafe fn create_buffer<T>(&self, length: usize) -> GPUResult<Buffer<T>> {
+        assert!(length > 0);
+        // This is the unsafe call, the rest of the function is safe code.
+        let buffer = DeviceBuffer::<u8>::uninitialized(length * std::mem::size_of::<T>())?;
+
+        Ok(Buffer::<T> {
+            buffer,
+            length,
+            _phantom: std::marker::PhantomData,
+        })
+    }
+
+    /// Creates a new buffer on the GPU and initializes with the given slice.
+    pub fn create_buffer_from_slice<T>(&self, slice: &[T]) -> GPUResult<Buffer<T>> {
+        // The number of bytes is used for the allocations.
+        let bytes_len = slice.len() * std::mem::size_of::<T>();
+
+        // Transmuting types is safe as long a sizes match.
+        let bytes = unsafe {
+            std::slice::from_raw_parts(slice.as_ptr() as *const T as *const u8, bytes_len)
+        };
+
+        // It is only unsafe as long as the buffer isn't initialized, but that's what we do next.
+        let mut buffer = unsafe { DeviceBuffer::<u8>::uninitialized(bytes_len)? };
+        // It is safe as we synchronize the stream after the call.
+        unsafe { buffer.async_copy_from(bytes, &self.stream)? };
```

Comment:
is this guranteed to intialize all bytes in `bytes`, independent of `self.stream`?

Reply:
I don't understand what you mean. It transfers all bytes from `bytes` into the memory allocated on the GPU via `buffer`. If the buffer is the same size as the bytes copied, all bytes are initialized.

Reply:
I didn‘t understand the code right, thanks all good